man/man8/gotlskeylog.8
.TH gotlskeylog 8  "2023-10-01" "USER COMMANDS"
.SH NAME
gotlskeylog \- Print TLS secrets of golang app for TLS decryption. Uses Linux eBPF/bcc.
.SH SYNOPSIS
.B gotlskeylog.py [-h] [-p PID] [-u UID] [--max-label-size MAX_LABEL_SIZE]
.B                [--max-client-random-size MAX_CLIENT_RANDOM_SIZE]
.B                [--max-secret-size MAX_SECRET_SIZE] [-o OUTPUT]
.B                path [path ...]
.SH DESCRIPTION
gotlskeylog logs TLS secrets from golang application in a SSLKEYLOGFILE format,
which then can be used in a Wireshark to decrypt TLS messages. It supports only
applications written in a Golang and compiled with recent versions of golang
that uses new register based ABI.

It works by tracing
"crypto/tls.(*Config).writeKeyLog" function and interception all 3 arguments of
this function. \fBNote\fP,
this tool currently only works on amd64 architecture.

For more information refer to this page:
.UR https://wiki.wireshark.org/TLS
.UE
and
.UR https://datatracker.ietf.org/doc/html/draft-thomson-tls-keylogfile-01
.UE
for SSLKEYLOGFILE spec.
Since this uses BPF, only the root user can use this tool.
.SH REQUIREMENTS
CONFIG_BPF and bcc.
.SH OPTIONS
.TP
\fB-h\fP, \fB--help\fP
Print help message and exit.
.TP
\fB-p\fP PID, \fB--pid\fP PID
Trace this PID only
.TP
\fB-u\fP UID, \fB--uid\fP UID
Trace this UID only
.TP
\fB--max-label-size\fP MAX_LABEL_SIZE
Max label size, should be in [32, 128] range, default is 32 as defined by spec.
.TP
\fB--max-client-random-size\fP MAX_CLIENT_RANDOM_SIZE
Max size of client_random, should be in [32, 256], default is 32 as defined by spec.
.TP
\fB--max-secret-size\fP MAX_SECRET_SIZE
Max secret size, should be in [48, 256] range, default is 64.
.TP
-\fBo\fP OUTPUT, \fB--output\fP OUTPUT
Output file to use instead of stdout
.TP
\fBpath\fP [path ...]
Path to application binaries, could be passed more than one to trace multiple
applications.
.SH EXAMPLES
.TP
Log TLS secrets from Golang app in /tmp/test path and output to stdout:
# 
.B gotlskeylog.py /tmp/test
.TP
You also can trace multiple golang programs at the same time:
# 
.B gotlskeylog.py /opt/bin/foo /opt/bin/bar
.TP
Output can be saved into file:
# 
.B gotlskeylog.py -o /tmp/sslkeylog.txt /tmp/test
.SH SOURCE
This is from bcc.
.IP
https://github.com/iovisor/bcc
.PP
Also look in the bcc distribution for a companion _examples.txt file containing
example usage, output, and commentary for this tool.
.SH OS
Linux
.SH STABILITY
Unstable - in development.
.SH SEE ALSO
wireshark(1) tcpdump(1) sslsniff(8)


debian/changelog
bcc (0.28.0-1) unstable; urgency=low

  * Support for kernel up to 6.3.
  * new libbpf tool: tcppktlat.
  * bcc tool updates: funcslower, wakeuptime, profile, offcputime, deadlock, funccount, argdist, kvmexit, runqlen and cpuunclaimed.
  * libbpf tool update: memleak, tcprtt, tcpconnlat, funclatency, syscount, cpufreq, biosnoop.
  * support ringbuf_query for bcc tools.
  * handle '[uprobes]' memory mapped file properly during stack tracing.
  * Fix maximum allowed index for print_linear_hist for bcc tools.
  * add module kfunc/kretfunc support.
  * clang rewriter: initialize only the requested parameters
  * filter with available_filter_functions to make multi-functions kprobes more robust for both bcc and libbpf tools.
  * doc update, other bug fixes and tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 28 Jun 2023 17:00:00 +0000

bcc (0.27.0-1) unstable; urgency=low

  * Support for kernel up to 6.2
  * bcc tool updates for ttysnoop, slabratetop, readahead, nfsslower, cpudist, cachetop, cachestat, etc.
  * libbpf-tools updates for mdflush, drsnoop, statsnoop, ttysnoop, softirqs, wakeuptime, cachestat, numamove, etc.
  * fix for incomplete static libraries
  * implement zip archive support
  * upgrade to use c++14 standard
  * new libbpf-tools: memleak
  * add loongarch support in libbpft-tools
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 02 Apr 2023 17:00:00 +0000

bcc (0.26.0-1) unstable; urgency=low

  * Support for kernel up to 6.1
  * bcc tool updates for biosnoop, opensnoop, biopattern, killsnoop, runqslower, offcputime, wakeuptime, etc.
  * libbpf-tools updates for klockstat, sigsnoop, hardirqs, softirqs, opensnoop, statsnoop, offcputime, tcplife, cpufreq, cpudist, etc.
  * new libbpf-tools: tcptop, tcpstates, biotop, capable
  * ci: add support for fedora 36 container and new workflow for containers
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 10 Aug 2022 17:00:00 +0000

bcc (0.25.0-1) unstable; urgency=low

  * Support for kernel up to 5.19
  * bcc tool updates for oomkill.py, biolatpcts.py, sslsniff.py, tcpaccept.py, etc.
  * libbpf tool updates for klockstat, opensnoop, tcpconnect, etc.
  * new bcc tools: tcpcong
  * new libbpf tools: tcpsynbl, mdflush, oomkill, sigsnoop
  * usdt: support xmm registers as args for x64
  * bpftool as a submodule now
  * remove uses of libbpf deprecated APIs
  * use new llvm pass manager
  * support cgroup filtering libbpf tools
  * fix shared lib module offset <-> global addr conversion
  * riscv support
  * LoongArch support
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 10 Aug 2022 17:00:00 +0000

bcc (0.24.0-1) unstable; urgency=low

  * Support for kernel up to 5.16
  * bcc tools: update for trace.py, sslsniff.py, tcptop.py, hardirqs.py, etc.
  * new libbpf tools: bashreadline
  * allow specify wakeup_events for perf buffer
  * support BPF_MAP_TYPE_{INODE, TASK}_STORAGE maps
  * remove all deprecated libbpf function usage
  * remove P4/B language support
  * major test infra change, using github actions now
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 14 Jan 2022 17:00:00 +0000

bcc (0.23.0-1) unstable; urgency=low

  * Support for kernel up to 5.15
  * bcc tools: update for kvmexit.py, tcpv4connect.py, cachetop.py, cachestat.py, etc.
  * libbpf tools: update for update for mountsnoop, ksnoop, gethostlatency, etc.
  * fix renaming of task_struct->state
  * get pid namespace properly for a number of tools
  * initial work for more libbpf utilization (less section names)
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 15 Nov 2021 17:00:00 +0000

bcc (0.22.0-1) unstable; urgency=low

  * Support for kernel up to 5.14
  * add ipv4/ipv6 filter support for tcp trace tools
  * add python interface to attach raw perf events
  * fix tcpstates for incorrect display of dport
  * new options for bcc tools runqslower, argdist
  * new libbpf-tools: filetop, exitsnoop, tcprtt
  * doc update, bug fixes and other tools improvement

 -- Yonghong Song <ys114321@gmail.com>  Wed, 15 Sep 2021 17:00:00 +0000

bcc (0.21.0-1) unstable; urgency=low

  * Support for kernel up to 5.13
  * support for debug information from libdebuginfod
  * finished support for map elements items_*_batch() APIs
  * add atomic_increment() API
  * support attach_func() and detach_func() in python
  * fix displaying PID instead of TID for many tools
  * new tools: kvmexit.py
  * new libbpf-tools: gethostlatency, statsnoop, fsdist and solisten
  * fix tools ttysnoop/readahead for newer kernels
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 16 Jul 2021 17:00:00 +0000

bcc (0.20.0-1) unstable; urgency=low

  * Support for kernel up to 5.12
  * Some basic support for MIPS
  * added bpf_map_lookup_batch and bpf_map_delete_batch support
  * tools/funclatency.py support nested or recursive functions
  * tools/biolatency.py can optionally print out average/total value
  * fix possible marco HAVE_BUILTIN_BSWAP redefine warning for kernel >= 5.10.
  * new tools: virtiostat
  * new libbpf-tools: ext4dist
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 5 May 2021 17:00:00 +0000

bcc (0.19.0-1) unstable; urgency=low

  * Support for kernel up to 5.11
  * allow BCC as a cmake subproject
  * add LPORT support in tcpconnlat and tcpconnect
  * added bpf_map_lookup_and_delete_batch support
  * new tools: virtiostat
  * new libbpf-tools: cpufreq, funclatency, cachestat
  * add install target to libbpf-tools
  * a few lua fixes
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 19 Mar 2021 17:00:00 +0000

bcc (0.18.0-1) unstable; urgency=low

  * Support for kernel up to 5.10
  * add bpf kfunc/kretfunc C++ example

tools/gotlskeylog.py
#!/usr/bin/env python
from __future__ import print_function
from bcc import BPF
import argparse
import binascii
import sys


def int_in_range(minval, maxval):
    def validator(val):
        try:
            intval = int(val)
        except ValueError:
            raise argparse.ArgumentTypeError("Incorrect int value %r" % val)
        else:
            if minval <= intval <= maxval:
                return intval
            else:
                raise argparse.ArgumentTypeError("Value should be in in range [%d, %d]" % (minval, maxval))

    return validator


class HelpFormatter(argparse.RawDescriptionHelpFormatter,
                    argparse.ArgumentDefaultsHelpFormatter):
    pass


examples = """examples:
    ./gotlskeylog /path/someapp                   # trace someapp and log TLS to stdout
    ./gotlskeylog -o /tmp/tls.log /path/someapp   # trace someapp and write keys into tls.log file
    ./gotlskeylog /path/foo /path/bar             # trace TLS keys from foo and bar Go apps
    ./gotlskeylog -p 181 someapp                  # trace only PID 181 of someapp
    ./gotlskeylog -u 1000 someapp                 # trace only UID 1000
"""

parser = argparse.ArgumentParser(
    description="Intercept TLS keys from Go app and write them in a SSLKEYLOGFILE format",
    formatter_class=HelpFormatter,
    epilog=examples
)
parser.add_argument("-p", "--pid", type=int, default=None, help="Trace this PID only")
parser.add_argument("-u", "--uid", type=int, default=None, help="Trace this UID only")
parser.add_argument("--max-label-size", type=int_in_range(32, 128), default=32,
                    help="Max label size, should be in [32, 128] range")
parser.add_argument("--max-client-random-size", type=int_in_range(32, 256),
                    default=32, help="Max size of client_random, should be in [32, 256]")
parser.add_argument("--max-secret-size", type=int_in_range(48, 256), default=64,
                    help="Max secret size, should be in [48, 256] range")
parser.add_argument("-o", "--output", type=argparse.FileType(mode="w"),
                    default=sys.stdout, help="Output file to use instead of stdout")
parser.add_argument("path", type=str, nargs='+')
args = parser.parse_args()


BPF_PROGRAM_TEMPLATE = """
#include <linux/ptrace.h>

// https://go.googlesource.com/go/+/refs/heads/dev.regabi/src/cmd/compile/internal-abi.md
#define GOLANG_REGS_PARM1(x) ((x)->ax)
#define GOLANG_REGS_PARM2(x) ((x)->bx)
#define GOLANG_REGS_PARM3(x) ((x)->cx)
#define GOLANG_REGS_PARM4(x) ((x)->di)
#define GOLANG_REGS_PARM5(x) ((x)->si)
#define GOLANG_REGS_PARM6(x) ((x)->r8)
#define GOLANG_REGS_PARM7(x) ((x)->r9)
#define GOLANG_REGS_PARM8(x) ((x)->r10)
#define GOLANG_REGS_PARM9(x) ((x)->r11)

#define MAX_LABEL_SIZE __MAX_LABEL_SIZE__
#define MAX_CLIENT_RANDOM_SIZE __MAX_CLIENT_RANDOM_SIZE__
#define MAX_SECRET_SIZE __MAX_SECRET_SIZE__

struct probe_KeyLogData {
    u64 timestamp_ns;
    u32 pid;
    u32 tid;
    u32 uid;
    u64 label_size;
    u64 client_random_size;
    u64 secret_size;
    char label[MAX_LABEL_SIZE];
    u8 client_random[MAX_CLIENT_RANDOM_SIZE];
    u8 secret[MAX_SECRET_SIZE];
};

BPF_PERCPU_ARRAY(keylog_data, struct probe_KeyLogData, 1);
BPF_PERF_OUTPUT(perf_KeylogOut);

int probe_WriteKeyLog(struct pt_regs *ctx) {
    u64 len;
    u64 addr;
    u32 zero = 0;
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u64 ts = bpf_ktime_get_ns();
    u32 pid = pid_tgid >> 32;
    u32 tid = pid_tgid;
    u32 uid = bpf_get_current_uid_gid();

    __PID_FILTER__
    __UID_FILTER__

    struct probe_KeyLogData *data = keylog_data.lookup(&zero);
    if (!data)
        return 0;

    data->timestamp_ns = ts;
    data->pid = pid;
    data->tid = tid;
    data->uid = uid;
    len = GOLANG_REGS_PARM3(ctx);
    data->label_size = len;
    addr = GOLANG_REGS_PARM2(ctx);
    bpf_probe_read_user(&data->label, min((size_t)len, (size_t)MAX_LABEL_SIZE), (char *)addr);
    // client random GOLANG_REGS_PARM4 GOLANG_REGS_PARM5 (rdi addr, rsi len, r8 cap)
    addr = GOLANG_REGS_PARM4(ctx);
    len= GOLANG_REGS_PARM5(ctx); /* parm6 is cap(clientRandom) */
    data->client_random_size = len;
    if (len <= MAX_CLIENT_RANDOM_SIZE) {
        bpf_probe_read_user(&data->client_random, len, (u8*)addr);
    }
    addr = GOLANG_REGS_PARM7(ctx);
    len= GOLANG_REGS_PARM8(ctx);
    data->secret_size = len;
    if (len <= MAX_SECRET_SIZE) {
        bpf_probe_read_user(&data->secret, len, (u8*)addr);
    }
    perf_KeylogOut.perf_submit(ctx, data, sizeof(struct probe_KeyLogData));
    return 0;
}
"""


def generate_program(args):
    prog = BPF_PROGRAM_TEMPLATE.replace('__MAX_LABEL_SIZE__', str(args.max_label_size))
    prog = prog.replace('__MAX_CLIENT_RANDOM_SIZE__', str(args.max_client_random_size))
    prog = prog.replace('__MAX_SECRET_SIZE__', str(args.max_secret_size))
    if args.pid is None:
        prog = prog.replace('__PID_FILTER__', '')
    else:
        prog = prog.replace('__PID_FILTER__', 'if (pid != %d) { return 0; }' % args.pid)

    if args.uid is None:
        prog = prog.replace('__UID_FILTER__', '')
    else:
        prog = prog.replace('__UID_FILTER__', 'if (uid != %d) { return 0; }' % args.uid)

    return prog


ebpf = BPF(text=generate_program(args))
for path in args.path:
    ebpf.attach_uprobe(name=path, sym="crypto/tls.(*Config).writeKeyLog",
        fn_name="probe_WriteKeyLog")


def print_event(cpu, data, size):
    event = ebpf["perf_KeylogOut"].event(data)
    label = event.label[:min(event.label_size, args.max_label_size)]
    client_random = event.client_random[:min(event.client_random_size, args.max_client_random_size)]
    secret = event.secret[:min(event.secret_size, args.max_secret_size)]

    label_truncated = event.label_size > args.max_label_size
    client_random_truncated = event.client_random_size > args.max_client_random_size
    secret_truncated = event.secret_size > args.max_secret_size

    # do not log truncated keys
    if not (label_truncated or client_random_truncated or secret_truncated):
        args.output.write("%s %s %s\n" % (
            label.decode("utf-8"),
            binascii.hexlify(bytearray(client_random)).decode("utf-8"),
            binascii.hexlify(bytearray(secret)).decode("utf-8"))
        )
        args.output.flush()
    else:
        # notify about truncated keys, so user could adjust buffer sizes
        truncated_msg = ["WARNING: (pid: %d, uid %d)" % (event.pid, event.uid)]
        if label_truncated:
            truncated_msg.append(" label truncated (size %d)" % event.label_size)
        if client_random_truncated:
            truncated_msg.append(" client_random truncated (size %d)" % event.client_random_size)
        if secret_truncated:
            truncated_msg.append(" secret truncated (size %d)" % event.secret_size)

        truncated_msg.append("\n")
        sys.stderr.write("".join(truncated_msg))


try:
    ebpf["perf_KeylogOut"].open_perf_buffer(print_event)
    while True:
        try:
            ebpf.perf_buffer_poll()
        except KeyboardInterrupt:
            exit()
finally:
    args.output.close()
